components/nimbus/src/enrollment.rs

```diff
@@ -601,30 +614,116 @@ impl<'a> EnrollmentsEvolver<'a> {
 
         let mut all_slugs = HashSet::with_capacity(existing_experiments.len());
         all_slugs.extend(existing_experiments.keys());
         all_slugs.extend(updated_experiments.keys());
-        all_slugs.extend(existing_enrollments.keys());
+        all_slugs.extend(existing_enrollment_map.keys());
+
+        // Step 1:
+        // make a hash map of feature_id to enrolled_experiment.slug
+        let mut locally_enrolled_feature_ids =
+            map_locally_enrolled_feature_ids(existing_enrollments, &existing_experiments);
 
         let mut updated_enrollments = Vec::with_capacity(all_slugs.len());
-        for slug in all_slugs {
-            let updated_enrollment = self.evolve_enrollment(
+
+        // Step 2:
+        // remove any feature-ids from the map that were being experimented
+        // on, but no longer are, according to the current updates.
+        for slug in all_slugs.clone() {
+            let _updated_enrollment = self.evolve_enrollment(
                 is_user_participating,
                 existing_experiments.get(slug).copied(),
                 updated_experiments.get(slug).copied(),
-                existing_enrollments.get(slug).copied(),
+                existing_enrollment_map.get(slug).copied(),
                 &mut enrollment_events,
             )?;
-            if let Some(enrollment) = updated_enrollment {
-                updated_enrollments.push(enrollment);
+            // if let Some(_enrollment) = updated_enrollment {
+            // What has changed? If we're still enrolled, do nothing.
+
+            // If not, remove the feature id - it's usable now.
+
+            // updated_enrollments.push(enrollment); XXX NOT NEEDED HERE, I
+            //if enrollment.status != EnrollmentStatus::Enrolled
+            // }
+        }
+
+        // Step 3: now we can add new experiments that avoid the features that
+        // already being experimented upon.
+
+        // the previous calls to evolve_enrollments will have prepopulated
+        // so...
+        enrollment_events.clear();
+
+        for slug in all_slugs {
+            let updated_experiment = updated_experiments.get(slug).copied();
+            if let Some(unwrapped_experiment) = updated_experiment {
+                let feature_id = &unwrapped_experiment.get_first_feature_id();
+
+                // If this feature_id is locally free, evolve the enrollment
+                // update the feature_id hashtable
+                if locally_enrolled_feature_ids.get(feature_id) == None {
+                    log::debug!("this feature is locally unused");
+
+                    let updated_enrollment = self.evolve_enrollment(
+                        is_user_participating,
+                        existing_experiments.get(slug).copied(),
+                        updated_experiments.get(slug).copied(),
+                        existing_enrollment_map.get(slug).copied(),
+                        &mut enrollment_events,
+                    )?;
```

Comment:
Just thinking this through: By this time, we know that this is we're in an ongoing experiment that we're not enrolled in.

1. ongoing experiment because it's in `updated_experiments`
2. if we were enrolled in it, then this feature id would be taken.

Reply:
(2) is important if we want want to call `evolve_enrollment` exactly once per experiment slug.